Add -num-nodes to e2e flags

akutz opened the pull request:
This patch introduces the `-num-nodes` flag to the invocation of `e2e.test` in `run_e2e.sh`. This is due to a [conversation](https://kubernetes.slack.com/archives/C78F00H99/p1551550250048900) in the \#k8s-conformance Slack channel.

Without this change the following conformance test fails:
* `[sig-apps] Daemon set [Serial] [It] should rollback without unnecessary restarts [Conformance]`

This test was introduced to `master` for the targeted 1.15 release - https://github.com/kubernetes/kubernetes/commit/8101b86ac4140636ceca52d87186dc1ecc07e438#diff-c944d1288edcaf37beebab811603bfd8.

**Test failure sans flag**
https://prow.k8s.io/view/gcs/k8s-conformance-cloud-provider-vsphere/head/ci/latest/1551630690

**Test success with flag**
https://gist.github.com/akutz/109a3c04d01a7b5c34af31f813807b2b

Please note that the image `akutz/kube-conformance` contains this patch.

dims commented:
related to https://github.com/kubernetes/kubernetes/issues/74853

neolit123 commented on run_e2e.sh:
```diff
@@ -42,13 +42,16 @@ ginkgo_args=(
 )
 
+# Number of nodes in the cluster.
+NUM_NODES="${NUM_NODES:-$(kubectl get nodes --no-headers | wc -l)}"
```
this will count control-plane nodes with `effect:NoSchedule` too.

akutz replied:
Thank you @neolit123! I realized why I thought `kubectl get nodes` returned only schedulable nodes in our Slack conversation -- sk8 doesn't deploy kubelets to control-plane only nodes. Those nodes run the API server, Controller Manager, and Scheduler as SystemD services without any kubelet involved.

It's may be accurate to then state that `kubectl get nodes` returns all registered kubelets.

akutz commented on run_e2e.sh:
```diff
@@ -42,13 +42,16 @@ ginkgo_args=(
 )
 
+# Number of nodes in the cluster.
+NUM_NODES="${NUM_NODES:-$(kubectl describe nodes | grep '^Taints:' | grep -vF 'node-role.kubernetes.io/master:NoSchedule' | wc -l)}"
```
Hi @neolit123,

Thank you for catching the previous error. Would you please take a look at this? Thank you!

neolit123 replied:
see https://github.com/kubernetes/kubernetes/pull/74854/files#diff-0efd26687540ba2cbfd9547019622551R63

akutz replied:
Per Slack, Lubomir noted:

> the list under `Taints:` can span on multiple lines.

akutz replied:
Thank you @neolit123.

akutz committed:
Add -num-nodes to e2e flags

This patch introduces the -num-nodes flag to the invocation of e2e.test
in run_e2e.sh. This is due to a conversation in the k8s-conformance
Slack channel -
https://kubernetes.slack.com/archives/C78F00H99/p1551550250048900.

timothysc commented:
This seems like a flailing in the tests TBH.  
/assign @stevesloka @johnSchnake

timothysc commented on run_e2e.sh:
```diff
@@ -42,13 +42,17 @@ ginkgo_args=(
 
+# Attempt to detect the number of nodes in the cluster on which workloads
+# may be scheduled.
+NUM_NODES="${NUM_NODES:-$(kubectl get nodes -o=jsonpath='{range .items[*]}{.spec.taints}{"\n"}{end}' | grep -cv 'effect:NoSchedule')}"
```
This should really be an initialization function of the framework vs. something external.

neolit123 replied:
tracking issue is here: https://github.com/kubernetes/kubernetes/issues/74853

akutz replied:
Hi @timothysc,

I agree 100%, and that's essentially what BenTheElder on Slack was saying as well. However, I encountered the problem while trying to fix broken conformance tests, and I fixed what I could on my end.

If you're fine with conformance against `ci/latest` failing until `e2e.test` has a sane default (it's currently `-1`) for `-num-nodes`, then I'm fine letting them fail as well.

johnSchnake commented:
Closing this PR; the whole repo should be archived as well as far as I know.
